File: julia/abc200_209/abc201_F.jl

using Random
infile = stdin
## Type Shortcuts (to save my wrists and fingers :))
const I = Int64; const VI = Vector{I}; const SI = Set{I}; const PI = NTuple{2,I};
const TI = NTuple{3,I}; const QI = NTuple{4,I}; const VPI = Vector{PI}; const SPI = Set{PI}
const VC = Vector{Char}; const VS = Vector{String}; VB = Vector{Bool}; VVI = Vector{Vector{Int64}}
const F = Float64; const VF = Vector{F}; const PF = NTuple{2,F}

gs()::String = rstrip(readline(infile))
gi()::Int64 = parse(Int64, gs())
gf()::Float64 = parse(Float64,gs())
gss()::Vector{String} = split(gs())
gis()::Vector{Int64} = [parse(Int64,x) for x in gss()]
gfs()::Vector{Float64} = [parse(Float64,x) for x in gss()]



struct segtree{T}; _n::I; sz::I; mylog::I; d::Vector{T}; op; e::T; end
function segtree{T}(n::I,op,e::T) where {T}
    sz::I = 1; mylog::I = 0; while sz < n; sz <<= 1; mylog += 1; end
    return segtree{T}(n,sz,mylog,fill(e,2*sz),op,e)
end
function segtree{T}(v::Vector{T},op,e::T) where {T}
    sz::I = 1; mylog::I = 0; while sz < length(v); sz <<= 1; mylog += 1; end
    d::Vector{T} = fill(e,2*sz)
    d[sz:sz+length(v)-1] .= v
    s::segtree{T} = segtree{T}(length(v),sz,mylog,d,op,e)
    for i in sz-1:-1:1; _segtreeUpdate(s,i); end
    return s
end
function segtreeSet(s::segtree{T},p::I,x::T) where {T}
    p += s.sz-1
    s.d[p] = x
    for i in 1:s.mylog; _segtreeUpdate(s,p >> i); end
end
function segtreeGet(s::segtree{T},p::I) where {T}
    p -= 1
    return s.d[s.sz+p]
end
## This version, unlike atcoder verison, returns op(a[l],a[l+1],...,a[r])
function segtreeProd(s::segtree{T},l::I,r::I) where {T}
    l += s.sz-1; r += s.sz-1+1 ## Add 1 for our version
    sml::T = s.e; smr::T = s.e
    while l < r
        if (l & 1 == 1); sml = s.op(sml,s.d[l]); l += 1; end
        if (r & 1 == 1); r -= 1; smr = s.op(s.d[r],smr); end
        l >>= 1; r >>= 1
    end
    return s.op(sml,smr)
end
function segtreeAllprod(s::segtree{T}) where {T}
    return s.d[1]
end
function _segtreeUpdate(s::segtree{T},k::I) where {T}
    if k > 0; s.d[k] = s.op(s.d[2*k],s.d[2*k+1]); end
end

function segtreeop(a::I,b::I)::I
    return a < b ? a : b
end

function solve(N::I,P::VI,A::VI,B::VI,C::VI)
    inf::I = 10^18
    ans::I = inf
    st = segtree{I}(N+1,segtreeop,inf)
    dp::VI = fill(inf,N)
    pos::VI = fill(0,N)
    for i in 1:N; pos[P[i]] = i; end
    Asum::VI = fill(0,N+1)
    Bsum::VI = fill(0,N+1)
    Csum::VI = fill(0,N+1)
    for i in 1:N; Asum[i+1] = Asum[i] + A[i]; end
    for i in 1:N; Bsum[i+1] = Bsum[i] + min(A[i],B[i]); end
    for i in 1:N; Csum[i+1] = Csum[i] + min(A[i],C[i]); end
    for i in 1:N
        dp[i] = min(Bsum[i],segtreeProd(st,1,pos[i])+Asum[i])
        ans = min(ans,dp[i]+Csum[N+1]-Csum[i+1])
        segtreeSet(st,pos[i],dp[i]-Asum[i+1])
    end
    return ans
end

function main(infn="")
    global infile
    infile = (infn != "") ? open(infn,"r") : length(ARGS) > 0 ? open(ARGS[1],"r") : stdin
    N::I = gi()
    P::VI = gis()
    A::VI = fill(0,N)
    B::VI = fill(0,N)
    C::VI = fill(0,N)
    for i in 1:N; A[i],B[i],C[i] = gis(); end
    ans = solve(N,P,A,B,C)
    println(ans)
end

Random.seed!(8675309)
main()

#using Profile, StatProfilerHTML
#Profile.clear()
#@profile main("B.in")
#Profile.clear()
#@profilehtml main("B.in")

File: python/abc200_209/abc201_F.py

import sys
infile = sys.stdin.buffer

def gs()  : return infile.readline().rstrip()
def gi()  : return int(gs())
def gf()  : return float(gs())
def gss() : return gs().split()
def gis() : return [int(x) for x in gss()]
def gfs() : return [float(x) for x in gss()]

class segtree :
    def __init__(self,n=1,op=sum,e=0,v=None) :
        if v is not None : n = len(v)
        self.n = n; self.sz = 1; self.log = 0; self.op=op; self.e=e
        while self.sz < n : self.sz *= 2; self.log += 1
        self.d = [self.e for i in range(2*self.sz)]
        if v is not None :
            for i in range(n) : self.d[self.sz+i] = v[i]
            for i in range(n-1,0,-1) : self._update(i)

    def _update(self,k) :
        self.d[k] = self.op(self.d[2*k],self.d[2*k+1])

    def set(self,p,x) :
        p += self.sz
        self.d[p] = x
        for i in range(1,self.log+1) : self._update(p>>i)

    def get(self,p,x) : return self.d[self.sz+p]

    def prod(self,l,r) :
        r += 1 ## want to get product from l to r inclusive
        sml = self.e; smr = self.e; l += self.sz; r += self.sz
        while (l < r) :
            if (l & 1) : sml = self.op(sml, self.d[l]); l += 1
            if (r & 1) : r -= 1; smr = self.op(self.d[r],smr)
            l >>= 1; r >>= 1
        return self.op(sml,smr)

    def allprod(self) : return self.d[1]

def main(infn="") :
    global infile
    infile = open(infn,"r") if infn else open(sys.argv[1],"r") if len(sys.argv) > 1 else sys.stdin
    N = gi()
    P = gis()
    A = [0] * N
    B = [0] * N
    C = [0] * N
    inf = 10**18
    for i in range(N) : A[i],B[i],C[i] = gis()
    pos = [0] * N
    for i in range(N) : pos[P[i]-1] = i
    minAB = [min(A[i],B[i]) for i in range(N)]
    minAC = [min(A[i],C[i]) for i in range(N)]
    cumMinAB = [0] * N; xx = 0
    for i in range(N) : xx += minAB[i]; cumMinAB[i] = xx
    cumMinAC = [0] * N; xx = 0
    for i in range(N) : xx += minAC[i]; cumMinAC[i] = xx
    cumA = [0] * N; xx = 0
    for i in range(N) : xx += A[i]; cumA[i] = xx
    st = segtree(N,min,inf,[inf]*N)
    dp = [inf] * N
    for i in range(N) :  ## Assume i is the rightmost fixed element
        v1 = 0 if i == 0 else cumMinAB[i-1] ## Case where we don't fix any elements to the left of i
        v2 = inf if pos[i] == 0 else st.prod(0,pos[i]-1) + cumA[i-1] ## Case where we also fix some element(s) to the left of i and use A_i moves to get elements in the right place
        dp[i] = min(v1,v2)
        st.set(pos[i],dp[i]-cumA[i])
    best = inf
    for i in range(N) :
        cand = dp[i] + cumMinAC[N-1] - cumMinAC[i]
        best = min(best,cand)
    print(best)
    ## Assume we fix element i as the rightmost fixed element.  Consider how much it costs to order the left side.  We have two choices.
    ## -- Don't fix any other elements to the left of i.  We then need to apply min(A_i,B_i) operations to all elements less than i
    ##    (easy to calculate, this is just a cumulative sum)
    ## -- Pick another element j left of i to fix.  This person must have a lower id than person i.  Our answer is then
    ##    dp[j] + sum from j+1 to i-1 (A_i) == sum from 1 to i-1 (A_i) + (dp[j] - sum from 1 to j (A_i))
    ##    The last element can be done with a segment tree

if __name__ == '__main__' :
    main()
    sys.stdout.flush()

